docs/errors.md

```diff
@@ -0,0 +1,124 @@
+
+
+## Exception.raiseInternal
+
+`Exception.raiseInternal` is for things that should not happen. Note: not for things
+where the user can trigger it by making a mistake - use a `RuntimeError` for that.
+This is for things where the user should not have been able to make this happen. It
+indicates a bug in our code, not in the user's code.
+
+For example, if we call a builtin function where the parameter type is
+`List<String>`, we will still need to go through the list and extract the F# `string`
+from the `DString`s in the list. Since the typechecker has already checked that the
+`DList` should only have `DString`s in it, finding anything else in there should be
+impossible and is a bug in the runtime.
+
+`Exception.raiseInternal`, in the cloud runtime, creates an exception that we can
+track and get alerts from Rollbar, including a stacktrace and a set of useful
+metadata for debugging the issue. We should debug and solve these.
+
+## Errors in Builtin functions
+
+When writing a builtin function, typically we handle errors by returning a
+`@darklang.stdlib.Result.Result` value.
+
+We have been in the habit of creating `Result<_, String>` - don't do that. Instead,
+create a darklang type for the error. That type will be created in a package even
+though it's used from a builtin - that's fine. Eg `Builtin.HttpClient.request`
+returns `@darklang.stdlib.HttpClient.Request`, which is `Result<_,
+@darklang.stdilib.HttpClient.Error>`.
+
+Types are cheap: it is better to create a new type that will be clear and obvious to
+users, rather than trying to reuse a general type or a string.
+
+### Raise RuntimeErrors for type errors in builtins
+
+If a builtin has a type error, such as the wrong type being returned by
+`Interpreter.applyFnVal`, use a `TypeChecker.Error` instead of a `Result`.
+
+If you expect a type, as indicated by the type signature, and it is not that type, a
+TypeChecker.Error is appropriate. Use `Result`s for errorable conditions like trying
+to read a file which doesn't exist.
+
+### F# exceptions in builtins
+
+If code raises a dotnet `System.Exception`-based Exception, it should be caught at
+the callsite and should not spread any further. Do not use the error message from
+dotnet exceptions darklang-specific error mechanisms.
+
+## RuntimeError implementation
+
+Unusually, a `RuntimeError` is actually a wrapper around a `Dval`, which corresponds
+to a type in `@darklang.languageTools.runtimeErrors`. Using a `Dval` here allows us
+work around some of the problems with the F# module ordering, and define/create
+`RuntimeError`s from anywhere.
+
+### Making a new RuntimeError
+
+To make a new `RuntimeError`, create a package type for it in
+`@darklang.languageTools.runtimeErrors`. Then you will need a set of functions in F#
+to produce it and convert it to a `Dval`, as well as handling the
+stringification in `@darklang.languageTools.runtimeErrors.error.toString`.
+
```

Comment:
Should we create a new module for the error in darklang.languageTools.runtimeErrors ? or should we add the error to the existing Error type ?

and by @darklang.languageTools.runtimeErrors.error.toString if you are referring to the  Error type could you please make the `e `uppercased so it is not confusing